### PR TITLE
Feature/boas 521 feature flags

### PR DESCRIPTION
**Pull request by francescobrenci-AR**
## Describe your changes

Basic implementation of Feature Flags - follows the pattern used in https://github.com/Planning-Inspectorate/applications-service

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [ ] I have performed a self-review of my own code
- [ ] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [ ] I have provided details on how I have tested my code
- [ ] I have referenced the ticket number above
- [ ] I have provided a description of the ticket
- [ ] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: apps/api/src/server/config/schema.js [hunk 1]
```diff
@@ -28,5 +28,6 @@ export default joi.object({
 		levelFile: joi.string(),
 		levelStdOut: joi.string()
 	}),
-	cwd: joi.string()
+	cwd: joi.string(),
+	featureFlags: joi.object()
```
Comment (mayya-bond-kc) [inline apps/api/src/server/config/schema.js:32]: Do we want to define here what keys we might expect in this object?

File: apps/document-storage/src/server/config/config.js [hunk 2]
```diff
@@ -18,7 +18,12 @@ const { value, error } = schema.validate({
 		levelStdOut: environment.LOG_LEVEL_STDOUT || 'debug'
 	},
 	cwd: url.fileURLToPath(new URL('..', import.meta.url)),
-	defaultApiVersion: '1'
+	defaultApiVersion: '1',
+	// flag name convention: featureFlag[ jira number ][ferature shoret description]
+	// set Feature Flag default val here [default: true] - will be overwritted by values cming from the .env file
+	featureFlags: {
+		featureFlagBoasXTestFeature: environment.FEATURE_FLAG_BOAS_X_TEST_FEATURE === 'true'
+	}
```
[the remark this answers is not in the document]
Comment (francescobrenci-AR) [reply apps/document-storage/src/server/config/config.js:26]: yes - each feature that needs a flag needs one setup here (default) and a matching one in Azure > app > config.
On our code we then use the featureFlags keys to retrieve the flag using `isFeatureActive(featureFlag[ jira number ][ferature shoret description])`

File: apps/api/src/server/config/schema.js (hunk 1 above)
Comment (francescobrenci-AR) [reply apps/api/src/server/config/schema.js:32]: The keys here can be any and as many as we need - is there a way to define this pattern in `joy`? they will all have a boolean as value

File: apps/web/src/common/feature-flags.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+export const isFeatureActive = (featureFlagName) => {
+	pino.info(`flag name: ${featureFlagName}`);
+
+	if (
+		!featureFlagName ||
+		!Object.prototype.hasOwnProperty.call(config.featureFlags, featureFlagName)
+	) {
+		pino.info(`a flag name must be supplied: ${featureFlagName} does not exist`);
```
Comment (lutamKC) [inline apps/web/src/common/feature-flags.js:17]: all web logs should start with [WEB]

File: apps/web/src/common/feature-flags.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import config from '@pins/web/environment/config.js';
```
Comment (lutamKC) [inline apps/web/src/common/feature-flags.js:1]: what's the reason for the new "common" folder? is it something that could affect client-side code?

Comment (francescobrenci-AR) [reply apps/web/src/common/feature-flags.js:1]: yes - you might want to put client side code behind a flag

File: apps/api/src/server/config/schema.js (hunk 1 above)
Comment (mayya-bond-kc) [reply apps/api/src/server/config/schema.js:32]: Yeah you would say: 
```
join.object({
    someFeature: joi.boolean()
})
```

Comment (francescobrenci-AR) [reply apps/api/src/server/config/schema.js:32]: that means every time we add a flag the schema needs updating though

Comment (francescobrenci-AR) [reply apps/api/src/server/config/schema.js:32]: I was looking for something more like this https://joi.dev/api/?v=17.7.0#objectpatternpattern-schema-options

File: apps/api/src/server/config/schema.js [hunk 5]
```diff
@@ -28,5 +28,6 @@ export default joi.object({
 		levelFile: joi.string(),
 		levelStdOut: joi.string()
 	}),
-	cwd: joi.string()
+	cwd: joi.string(),
+	featureFlags: joi.object().pattern(/featureFlagBoas\d+[A-Za-z]+/, joi.boolean())
```
Comment (mayya-bond-kc) [inline apps/api/src/server/config/schema.js:32]: Ah cool cool, so all feature flags will start with the same set of characters

**Review by mayya-bond-kc — APPROVED** [review]

**Review by lutamKC — APPROVED** [review]

**Pull request merged**